### ISSUE TITLE
Feature: Enforce loading via HTTPS (redirect)

### ISSUE THREAD

**Issue by mingsai** [mevdschee/php-crud-api#676]
## Background
I have been able to successfully enforce HTTPS redirection for the file by adding a one liner after the first namespace that requires a php file that evaluates the headers and forces the php to load over HTTPS. I'm not sure whether this is something that ought to be part of the main script or left outside of the domain. 

Other implementations I considered would be to do the redirection within the customization middleware, but I am not yet certain how that interaction would impact ongoing events. At this time having the call to redirect as high up in the script as possible seemed to be the cleanest approach. 

I also tried other things such as configuring an HTTP access file (.htaccess) but I found that my endpoints could still be accessible via HTTP (which led me to consider that the security of the script is at risk depending on the complex interaction of various environments). The only way to fully ensure that the script loads over HTTPS seems to me is to embed the logic within the actual api.php file so that when loading the script would redirect to the HTTPS itself.

I do not believe that anything would be better than having this functionality integrated fully and configurable via a switch. However I understand that some scope issues related to this must be considered. So below are some considerations and the code that one may use to make the best determination.

**Pros**
- Securing the middleware is a really important piece of implementing in a production environment
- Embedding this feature within a configurable variable leads to better code and consistency
- With multiple authorization and validation handlers this might be within scope 

**Cons**
- Technically, the implementation details may be beyond the domain or scope of the key focus of the project 

### Here's the code I  used to reload the script onto an HTTPS channel (via StackOverflow).
```
<?php
function forceHTTPS(){
  $httpsURL = 'https://'.$_SERVER['HTTP_HOST'].$_SERVER['REQUEST_URI'];
  if( count( $_POST )>0 )
    die( 'Page should be accessed with HTTPS, but a POST Submission has been sent here. Adjust the form to point to '.$httpsURL );
  if( !isset( $_SERVER['HTTPS'] ) || $_SERVER['HTTPS']!=='on' ){
    if( !headers_sent() ){
      header( "Status: 301 Moved Permanently" );
      header( "Location: $httpsURL" );
      exit();
    }else{
      die( '<script type="javascript">document.location.href="'.$httpsURL.'";</script>' );
    }
  }
}
forceHTTPS();
?>
// at some point in the top of the api.php I simply:
namespace ...
require('secret-force-https-file.php');
```

**Comment by mevdschee** [comment]
>I have been able to successfully enforce HTTPS redirection

Why is your API available over HTTP anyway? HTTPS is the new standard.

>I also tried other things such as configuring an HTTP access file (.htaccess)

It is certainly possible with a htaccess file:

        # redirect to https
        RewriteEngine on
        RewriteCond %{HTTPS} off
        RewriteRule ^(.*)$ https://%{HTTP_HOST}$1 [R=301,L]

see: https://stackoverflow.com/questions/1108706/correctly-switching-between-http-and-https-using-htaccess

>Other implementations I considered would be to do the redirection within the customization middleware,

That is also possible. It would be fairly easy to write a "SslRedirectMiddleware" and bundle it with the software.

>I do not believe that anything would be better than having this functionality integrated fully and configurable via a switch. 

Serving content without HTTPS seems to be configuration option in the web server and not of the application, so as you have already identified, this may be out of scope for application software.

I'll keep this open as a feature request for now.

**Comment by mingsai** [comment]
The redirect was not working in my environment (too many redirects because I have other .htaccess enabled throughout the site and getting the right mix of behaviors was problematic). This is because the server will act on .htaccess in an accumulative fashion.

I think this would work best if the behavior were switchable. There are some production environments, such as those on intranets where HTTPS may not be enforced so giving users the option of enforcement with a default to ON seems like a good option. 

I'll experiment with using existing handlers to create the needed effect. Not sure how to make an SslRedirectMiddleware so I'll take another deep dive into the codebase.

Thanks again.

**Comment by mevdschee** [comment]
>Not sure how to make an SslRedirectMiddleware

I have added an example, check it out!

**Comment by mingsai** [comment]
Thank you for sharing the example illustrating how to create the middleware. If one were going to integrate the example then it would be added to the SimpleRouter switch case, however when I look at it and consider the overall objective (allow users to enforce HTTPS), I come up with the following observations:

- Going with a middleware seems a clean way to integrate the functionality (SoC)
- Middleware gives us a closure but there is nothing within that scope which needs doing 
- Once the middleware is run, the redirect occurs

- A property on the config similar to `debug` and `basePath` might be a better way to handle the redirection since we are looking solely to toggle a behavior
- The property approach might be better if linked to the initialization cycle of Config in the Api call
- Maybe the middleware process can be added as a private method to Config and executed based on the new property?

**Questions I am exploring on the use of a new property on Config**
Should the redirect method be on Config or another object (eg. SimpleRouter or Api)?
What's the overall impact (amount of code to be changed)?
Is the result effective and the simplest possible thing we could do?

Given these questions, I think I'd like to explore the simple alternative of adding a property which we could toggle for the forced redirection. 

For example SimpleRouter currently injects two properties from the Config object during initialization. 

- Why not add a third property and modify the Dependency Injection to take the Config object and parse out the requisite properties for the appropriate behavior? 
- If passing security info is an issue then maybe coalesce the properties into a single parameter with all of the config related properties in one object?

**Comment by mevdschee** [comment]
>I found that the current redirect is taking me to a my root

Are you sure? The test does not show this behavior.

**Comment by mevdschee** [comment]
>Maybe in the middleware we can simply pass an argument of true/false to have the desired behavior. 

Why would you enable the middleware and then disable it's behavior?

**Comment by mingsai** [comment]
My experience has shown me that for every rule there is  an exception. 

> Okay, if you need it, I guess it's fine, but why would you enable the middleware and then disable it's behavior?

My experience has conditioned me to believe that for every rule there is an exception. I believe we should allow developers to exercise that control. My specific case here would be to enable the specification by default and the toggle allows developers to override the behavior for some unknown future reason. 

Earlier I mentioned servers that for whatever reason may not have the proper SSL enabled. I guess that's the long answer. Short answer - edge cases.

**Comment by mingsai** [comment]
> > I found that the current redirect is taking me to a my root
> 
> Are you sure? The test does not show this behavior.

I noted earlier that I have other redirection happening in the environment. The redirect to the path default/index.php is occurring. The path I want is at `$_SERVER['SCRIPT_URI']` so I am looking at how we can redirect to the https version of that using the request object in this context.

**Comment by mevdschee** [comment]
@mingsai (off-topic) Your github profile has an web url with a typo. It says 'www.tpmmiecarter.com' instead of 'www.tommiecarter.com'.

**Comment by mevdschee** [comment]
I have released a new version (v2.9.4) with the SSL redirect middleware included. I guess that makes this issue resolved. Closing for now.
